From df3f90e3079b100aaf7e63286789bbefa74c0e55 Mon Sep 17 00:00:00 2001
From: Robin Gong <b38343@freescale.com>
Date: Fri, 25 Dec 2015 10:06:27 +0800
Subject: [PATCH 0798/5242] MLK-12076-3: dma: imx-sdma: add imx6ul device type

commit  618285375d780b6ea8edd227c50dbe363ee206a8 from
https://source.codeaurora.org/external/imx/linux-imx.git

Add imx6ul device type to enable the ERR008517 workaround or not by dts.

Signed-off-by: Robin Gong <b38343@freescale.com>
(cherry picked from commit 17d6a090b7a39bfd7836a3685d20201dcb0fa25e)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/dma/imx-sdma.c |   22 +++++++++++++++++++---
 1 file changed, 19 insertions(+), 3 deletions(-)

diff --git a/drivers/dma/imx-sdma.c b/drivers/dma/imx-sdma.c
index 1831a78..8e021ba 100644
--- a/drivers/dma/imx-sdma.c
+++ b/drivers/dma/imx-sdma.c
@@ -520,6 +520,12 @@ struct sdma_engine {
 	.script_addrs = &sdma_script_imx6sx,
 };
 
+static struct sdma_driver_data sdma_imx6ul = {
+	.chnenbl0 = SDMA_CHNENBL0_IMX35,
+	.num_events = 48,
+	.script_addrs = &sdma_script_imx6sx,
+};
+
 static struct sdma_script_start_addrs sdma_script_imx7d = {
 	.ap_2_ap_addr = 644,
 	.uart_2_mcu_addr = 819,
@@ -570,6 +576,7 @@ struct sdma_engine {
 MODULE_DEVICE_TABLE(platform, sdma_devtypes);
 
 static const struct of_device_id sdma_dt_ids[] = {
+	{ .compatible = "fsl,imx6ul-sdma", .data = &sdma_imx6ul, },
 	{ .compatible = "fsl,imx6sx-sdma", .data = &sdma_imx6sx, },
 	{ .compatible = "fsl,imx6q-sdma", .data = &sdma_imx6q, },
 	{ .compatible = "fsl,imx53-sdma", .data = &sdma_imx53, },
@@ -1157,8 +1164,15 @@ static int sdma_config_channel(struct dma_chan *chan)
 			if (sdmac->peripheral_type == IMX_DMATYPE_ASRC_SP ||
 			    sdmac->peripheral_type == IMX_DMATYPE_ASRC)
 				sdma_set_watermarklevel_for_p2p(sdmac);
-		} else
+		} else {
+			/* ERR008517 fixed on i.mx6ul, no workaround needed */
+			if (sdmac->peripheral_type == IMX_DMATYPE_CSPI &&
+			    sdmac->direction == DMA_MEM_TO_DEV &&
+			    sdmac->sdma->drvdata == &sdma_imx6ul)
+				__set_bit(31, &sdmac->watermark_level);
+
 			__set_bit(sdmac->event_id0, sdmac->event_mask);
+		}
 
 		/* Address */
 		sdmac->shp_addr = sdmac->per_address;
@@ -2217,7 +2231,8 @@ static int sdma_suspend(struct device *dev)
 	int i, ret = 0;
 
 	/* Do nothing if not i.MX6SX or i.MX7D*/
-	if (sdma->drvdata != &sdma_imx6sx && sdma->drvdata != &sdma_imx7d)
+	if (sdma->drvdata != &sdma_imx6sx && sdma->drvdata != &sdma_imx7d
+	    && sdma->drvdata != &sdma_imx6ul)
 		return 0;
 
 	clk_enable(sdma->clk_ipg);
@@ -2255,7 +2270,8 @@ static int sdma_resume(struct device *dev)
 	int i, ret;
 
 	/* Do nothing if not i.MX6SX or i.MX7D*/
-	if (sdma->drvdata != &sdma_imx6sx && sdma->drvdata != &sdma_imx7d)
+	if (sdma->drvdata != &sdma_imx6sx && sdma->drvdata != &sdma_imx7d
+	    && sdma->drvdata != &sdma_imx6ul)
 		return 0;
 
 	clk_enable(sdma->clk_ipg);
-- 
1.7.9.5

